activestorage/test/service/shared_service_tests.rb

```diff
@@ -13,10 +13,17 @@ module ActiveStorage::Service::SharedServiceTests
       @key = SecureRandom.base58(24)
       @service = self.class.const_get(:SERVICE)
       @service.upload @key, StringIO.new(FIXTURE_DATA)
+
+      if self.class.const_defined?(:PUBLIC_SERVICE)
+        @public_service = self.class.const_get(:PUBLIC_SERVICE)
```

Comment:
Does it make sense to have separate test cases for public service tests? That seems clearer to me

Reply:
I'm not quite sure what you mean by this. Do you mean moving public service tests into a separate file (e.g. `shared_public_service_tests.rb`)?

Reply:
I mean keeping the shared service tests focused on `SERVICE` and creating separate test cases for each public version of a service (each with one `test "URL generation"` test). This way test cases are always testing just one service.

Reply:
I'm not sure how that could be cleaner. From what I understand, the logic for setup and upload a file would be moved into the `test "public URL generation"` test cases? That seems like more logic than just testing generating URL and leaving all the setup/uploading in the `setup` block.

I also wonder if it would make sense to run every test in `SharedServiceTests` for the public service, since we can't assume that a public service would behave similarly to a private service (even though it should for everything except URL generation)? Or would this be too overkill?

Reply:
> I'm not sure how that could be cleaner. From what I understand, the logic for setup and upload a file would be moved into the `test "public URL generation"` test cases?

To be more specific, this is what I would like to see, but applied all service tests:

```
# frozen_string_literal: true

require "service/shared_service_tests"
require "net/http"

if SERVICE_CONFIGURATIONS[:gcs_public]
  class ActiveStorage::Service::GCSPublicServiceTest < ActiveSupport::TestCase
    SERVICE = ActiveStorage::Service.configure(:gcs_public, SERVICE_CONFIGURATIONS)

    include ActiveStorage::Service::SharedServiceTests

    test "name" do
      assert_equal :gcs_public, @service.name
    end

    test "public URL generation" do
      url = @service.url(@key, filename: ActiveStorage::Filename.new("avatar.png"))

      assert_match(/storage\.googleapis\.com\/.*\/#{@key}/, url)

      response = Net::HTTP.get_response(URI(url))
      assert_equal "200", response.code
    end
  end
else
  puts "Skipping GCS Public Service tests because no GCS configuration was supplied"
end
```

This way

- There is less complexity in shared service setup blocks
- The public service contract is easier to understand because it has a separate test case
- Public services gain better test coverage (tested against all test methods in `ActiveStorage::Service::SharedServiceTests`)
- We aren't assuming that if `:gcs` config exists then the `:gcs_public` config exists too

Reply:
That makes sense, thanks for clarifying! I've updated it in the latest commit.